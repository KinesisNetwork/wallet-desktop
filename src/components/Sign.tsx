import * as copy from 'copy-to-clipboard'
import { Keypair } from 'js-kinesis-sdk'
import { kebabCase, startCase } from 'lodash'
import * as React from 'react'

import { InputField } from '@components/InputField'
import { SignProps } from '@containers/Sign'
import { SignTransactionForm } from '@containers/SignTransactionForm'
import { enumStringValues } from '@helpers/enumStringValues'
import { InputError, WalletLockError } from '@helpers/errors'
import { SignBehaviour, SignedMessage } from '@types'

export class Sign extends React.Component<SignProps> {
  displayName: string = 'Sign'

  tabs = () => {
    const behaviourOpts = enumStringValues(SignBehaviour)
    return behaviourOpts.map(b => {
      const behaviour = b as SignBehaviour
      return (
        <li
          key={behaviour}
          className={`${this.props.focus === behaviour && 'is-active'}`}
          onClick={() => this.props.changeSignFocus(behaviour)}
        >
          <a>
            <span>{b}</span>
          </a>
        </li>
      )
    })
  }

  render() {
    return (
      <div>
        <div className="tabs">
          <ul>{this.tabs()}</ul>
        </div>
        {this.props.focus === SignBehaviour.sign && <SignForm {...this.props} />}
        {this.props.focus === SignBehaviour.verify && <VerifyForm {...this.props} />}
        {this.props.focus === SignBehaviour.signTransaction && <SignTransactionForm />}
      </div>
    )
  }
}

interface State {
  copied: boolean
}
export class SignForm extends React.Component<SignProps, State> {
  displayName: string = 'SignForm'

  state: State = {
    copied: false,
  }
  signData = (ev: React.FormEvent<HTMLFormElement>) => {
    ev.preventDefault()
    try {
      this.validateProps()
      const secret = this.props.decryptedPrivateKey
      const kp = Keypair.fromSecret(secret)
      const signature = kp.sign(Buffer.from(this.props.signData.message, 'utf8'))
      this.props.signMessage(signature.toString('hex'))
    } catch (e) {
      this.props.callFormAlert({
        message: e.message,
        key: e.key
      })
    }
  }

  checkWalletIsUnlocked = () => {
    if (!this.props.isWalletUnlocked) {
      throw new WalletLockError()
    }
  }

  validateProps = (): void | never => {
    this.checkWalletIsUnlocked()
    this.checkValidEntry('message')
  }

  checkValidEntry = (key: keyof SignedMessage) => {
    if (!this.props.signData[key]) {
      throw new InputError(`${startCase(key)} is required`, `signdata-${kebabCase(key)}`)
    }
  }

  copySignature = () => {
    copy(this.props.signature)
    this.setState({ copied: true })
    setTimeout(() => this.setState({ copied: false }), 2000)
  }

  render() {
    const { signData, handleSignFormChange, signature } = this.props
    const { copied } = this.state
    return (
      <div className="is-centered">
        <form onSubmit={this.signData}>
          <InputField
            label="Message"
            value={signData.message}
            id="signdata-message"
            helpText="Enter the text to sign"
            onChangeHandler={newValue => handleSignFormChange('message', newValue)}
          />
          <div className="field is-grouped">
            <div className="control is-expanded">
              <button className="button is-fullwidth" type="submit">
                Sign
              </button>
            </div>
          </div>
          <div className="field">
            <label className="label is-small">Signature</label>
            <div className="field is-grouped">
              <div className="control is-expanded">
                <input className="input" value={signature} readOnly={true} disabled={true} />
              </div>
              <div className="control">
                <button
                  type="button"
                  className="button"
                  onClick={this.copySignature}
                  disabled={!signature}
                >
                  <span>{copied ? 'Copied' : 'Copy'}</span>
                  {copied && (
                    <span className="icon has-text-success">
                      <i className="fas fa-check" />
                    </span>
                  )}
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    )
  }
}

export class VerifyForm extends React.Component<SignProps> {
  displayName: string = 'VerifyForm'

  constructor(props: SignProps) {
    super(props)
  }

  verifyData = (ev: React.FormEvent<HTMLFormElement>) => {
    ev.preventDefault()
    if (!this.isValidForm()) {
      return
    }
    this.props.messageVerificationResult(this.verifyMessageSignature())
  }

  verifyMessageSignature = () => {
    try {
      const keypair = Keypair.fromPublicKey(this.props.verifyData.publicKey)
      return keypair.verify(
        Buffer.from(this.props.verifyData.message, 'utf8'),
        Buffer.from(this.props.verifyData.signature, 'hex'),
      )
    } catch (e) {
      return false
    }
  }

  isValidForm = (): boolean => {
    try {
      this.checkValidEntry('message')
      this.checkValidEntry('signature')
      this.checkValidEntry('publicKey')
      return true
    } catch (e) {
      this.props.callFormAlert({
        message: e.message,
        key: e.key
      })
      return false
    }
  }

  checkValidEntry = (key: keyof SignedMessage) => {
    if (!this.props.verifyData[key]) {
      throw new InputError(`${startCase(key)} is required`, `verify-${kebabCase(key)}`)
    }
  }

  render() {
    const { verifyData, handleVerifyFormChange } = this.props
    return (
      <div className="is-centered">
        <form onSubmit={this.verifyData}>
          <InputField
            label="Message"
            value={verifyData.message}
            id="verify-message"
            helpText="Enter the raw message"
            onChangeHandler={newValue => handleVerifyFormChange('message', newValue)}
          />
          <InputField
            label="Signature"
            value={verifyData.signature}
            id="verify-signature"
            helpText="Enter the signed dataset"
            onChangeHandler={newValue => handleVerifyFormChange('signature', newValue)}
          />
          <InputField
            label="Public Key"
            value={verifyData.publicKey}
            id="verify-public-key"
            helpText="Enter the public key of the claimed signer"
            onChangeHandler={newValue => handleVerifyFormChange('publicKey', newValue)}
          />
          <div className="field is-grouped">
            <div className="control is-expanded">
              <button className="button is-fullwidth" type="submit">
                Verify
              </button>
            </div>
          </div>
        </form>
      </div>
    )
  }
}
