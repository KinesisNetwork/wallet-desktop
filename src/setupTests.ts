import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
