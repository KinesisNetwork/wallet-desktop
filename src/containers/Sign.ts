import { connect } from 'react-redux'

import {
  changeSignFocus,
  invalidForm,
  messageVerificationResult,
  signMessage,
  updateSignForm,
  updateVerifyForm,
} from '@actions'
import { Sign as SignPresentation } from '@components/Sign'
import { Dispatch, RootState } from '@store'
import { FormAlert, RawMessage, SignBehaviour, SignedMessage, Wallet } from '@types'

const mapStateToProps = ({ sign, wallets, passwords, accounts }: RootState) => {
  const activeWallet = wallets.activeWallet as Wallet
  const isWalletUnlocked = accounts.accountsMap[activeWallet.publicKey].isUnlocked
  const decryptedPrivateKey = isWalletUnlocked
    ? passwords.livePasswords[activeWallet.publicKey].privateKey
    : ''
  return {
    decryptedPrivateKey,
    isWalletUnlocked,
    focus: sign.focus,
    signature: sign.signature,
    isValidSignature: sign.isValidSignature,
    signData: sign.signData,
    verifyData: sign.verifyData,
  }
}

const mapDispatchToProps = (dispatch: Dispatch) => ({
  changeSignFocus: (focus: SignBehaviour): void => {
    dispatch(changeSignFocus(focus))
  },
  handleSignFormChange: (field: keyof RawMessage, newValue: string) => {
    dispatch(updateSignForm({ field, newValue }))
  },
  handleVerifyFormChange: (field: keyof SignedMessage, newValue: string) => {
    dispatch(updateVerifyForm({ field, newValue }))
  },
  signMessage: (signature: string) => dispatch(signMessage(signature)),
  messageVerificationResult: (isValid: boolean) => dispatch(messageVerificationResult(isValid)),
  formIsInvalid: (error: FormAlert) => dispatch(invalidForm(error))
})

export type SignProps = ReturnType<typeof mapStateToProps> & ReturnType<typeof mapDispatchToProps>
export const Sign = connect(
  mapStateToProps,
  mapDispatchToProps,
)(SignPresentation)
