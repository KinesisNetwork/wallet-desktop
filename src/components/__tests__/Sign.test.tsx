import * as copy from 'copy-to-clipboard'
import { shallow } from 'enzyme';
import { Keypair } from 'js-kinesis-sdk'
import * as React from 'react'
import * as sinon from 'sinon'

import { Sign, SignForm } from '@components/Sign'
import '../../setupTests'

describe('Sign', () => {
  let props

  beforeEach(() => {
    props = {
      changeSignFocus: () => null,
      decryptedPrivateKey: '',
      focus: 'Sign',
      handleSignFormChange: () => null,
      handleSignVerifyFormChange: () => null,
      isValidSignature: false,
      isWalletUnlocked: false,
      messageVerificationResult: () => null,
      signData: {
        message: ''
      },
      signMessage: () => null,
      signature: '',
      verifyData: {
        message: '',
        publicKey: '',
        signature: ''
      }
    }
  })

  it('should render correctly', () => {
    const wrapper = shallow(<Sign {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders three tabs', () => {
    const wrapper = shallow(<Sign {...props} />)

    const tabs = wrapper.find('.tabs')
    const tab = tabs.find('li')

    expect(tab).toHaveLength(3)
    expect(tab.at(0).hasClass('is-active')).toBe(true)
    expect(tab.at(0).text()).toEqual('Sign')
    expect(tab.at(1).text()).toEqual('Verify')
    expect(tab.at(2).text()).toEqual('Sign Transaction')
  })

  describe('depending on the focus prop', () => {
    it('renders SignForm component', () => {
      const wrapper = shallow(<Sign {...props} />)

      expect(wrapper.find('SignForm')).toHaveLength(1)
    })

    it('renders VerifyForm', () => {
      const wrapper = shallow(<Sign {...props} focus='Verify' />)

      expect(wrapper.find('VerifyForm')).toHaveLength(1)
    })

    it('renders SignTransactionForm', () => {
      const wrapper = shallow(<Sign {...props} focus='Sign Transaction' />)

      expect(wrapper.find('Connect(SignTransactionForm)')).toHaveLength(1)
    })
  })
})

describe('SignForm', () => {
  let props

  beforeEach(() => {
    props = {
      callFormAlert: () => null,
      changeSignFocus: () => null,
      decryptedPrivateKey: '',
      focus: 'Sign',
      handleSignFormChange: () => null,
      handleSignVerifyFormChange: () => null,
      isValidSignature: false,
      isWalletUnlocked: false,
      messageVerificationResult: () => null,
      signData: {
        message: ''
      },
      signMessage: () => null,
      signature: '',
      verifyData: {
        message: '',
        publicKey: '',
        signature: ''
      }
    }
  })

  it('renders correctly', () => {
    const wrapper = shallow(<SignForm {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('renders a form with an InputField, a Sign button and a Signature field with a Copy button', () => {
    const wrapper = shallow(<SignForm {...props} />)

    const inputField = wrapper.find('InputField')
    const button = wrapper.find('button')

    expect(wrapper.find('form')).toHaveLength(1)
    expect(inputField).toHaveLength(1)
    expect(inputField.prop('label')).toEqual('Message')
    expect(inputField.prop('helpText')).toEqual('Enter the text to sign')
    expect(button).toHaveLength(2)
    expect(button.at(0).text()).toEqual('Sign')
    expect(button.at(1).text()).toEqual('Copy')
  })

  describe('calls callFormAlert', () => {
    it(' if the wallet is locked and the form is submitted', () => {
      const callFormAlertMock = jest.fn()
      const formAlertParam = {
        key: 'wallet-unlock-password',
        message: 'Wallet must be unlocked'
      }
      const wrapper = shallow(<SignForm {...props} callFormAlert={callFormAlertMock} />)

      const form = wrapper.find('form')
      form.simulate('submit', { preventDefault: () => null })

      expect(callFormAlertMock).toHaveBeenCalledWith(formAlertParam)
    })

    it('if the wallet is unlocked but message field is empty', () => {
      const callFormAlertMock = jest.fn()
      const formAlertParam = {
        key: 'signdata-message',
        message: 'Message is required'
      }
      const wrapper = shallow(
        <SignForm
          {...props}
          isWalletUnlocked={true}
          callFormAlert={callFormAlertMock}
        />
      )

      const form = wrapper.find('form')
      form.simulate('submit', { preventDefault: () => null })

      expect(callFormAlertMock).toHaveBeenCalledWith(formAlertParam)
    })
  })

  describe('if wallet is unlocked and a valid message is given', () => {
    it('calls the signMessage method', () => {
      const signMessageMock = jest.fn()
      const modifiedProps = {
        decryptedPrivateKey: '123',
        isWalletUnlocked: true,
        signData: {
          message: 'My Message'
        }
      }
      const wrapper = shallow(
        <SignForm
          {...props}
          {...modifiedProps}
          signMessage={signMessageMock}
        />
      )
      const keypairSpy = jest
        .spyOn(Keypair, 'fromSecret')
        .mockImplementation(() => ({ sign: () => 'qwe' }))

      const form = wrapper.find('form')
      form.simulate('submit', { preventDefault: () => null })

      expect(signMessageMock).toHaveBeenCalled()

      keypairSpy.mockReset()
    })

    it('displays the signature in the signature input field and enables the copy button', () => {
      const wrapper = shallow(
        <SignForm
          {...props}
          signature='asdf'
        />
      )

      const signatureLabel = wrapper.find('.label.is-small')
      const signatureInput = wrapper.find('input')
      const copyButton = wrapper.find('button').at(1)

      expect(signatureLabel).toHaveLength(1)
      expect(signatureLabel.text()).toEqual('Signature')
      expect(signatureInput).toHaveLength(1)
      expect(signatureInput.prop('value')).toEqual('asdf')
      expect(signatureInput.prop('readOnly')).toEqual(true)
      expect(signatureInput.prop('disabled')).toEqual(true)
      expect(copyButton.prop('disabled')).toEqual(false)
    })
  })
})
