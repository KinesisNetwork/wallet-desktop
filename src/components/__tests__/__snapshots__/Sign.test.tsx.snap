// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign should render correctly 1`] = `
<div>
  <div
    className="tabs"
  >
    <ul>
      <li
        className="is-active"
        key="Sign"
        onClick={[Function]}
      >
        <a>
          <span>
            Sign
          </span>
        </a>
      </li>
      <li
        className="false"
        key="Verify"
        onClick={[Function]}
      >
        <a>
          <span>
            Verify
          </span>
        </a>
      </li>
      <li
        className="false"
        key="Sign Transaction"
        onClick={[Function]}
      >
        <a>
          <span>
            Sign Transaction
          </span>
        </a>
      </li>
    </ul>
  </div>
  <SignForm
    changeSignFocus={[Function]}
    decryptedPrivateKey=""
    focus="Sign"
    handleSignFormChange={[Function]}
    handleSignVerifyFormChange={[Function]}
    isValidSignature={false}
    isWalletUnlocked={false}
    messageVerificationResult={[Function]}
    signData={
      Object {
        "message": "",
      }
    }
    signMessage={[Function]}
    signature=""
    verifyData={
      Object {
        "message": "",
        "publicKey": "",
        "signature": "",
      }
    }
  />
</div>
`;

exports[`SignForm renders correctly 1`] = `
<div
  className="is-centered"
>
  <form
    onSubmit={[Function]}
  >
    <InputField
      helpText="Enter the text to sign"
      id="signdata-message"
      label="Message"
      onChangeHandler={[Function]}
      value=""
    />
    <div
      className="field is-grouped"
    >
      <div
        className="control is-expanded"
      >
        <button
          className="button is-fullwidth"
          type="submit"
        >
          Sign
        </button>
      </div>
    </div>
    <div
      className="field"
    >
      <label
        className="label is-small"
      >
        Signature
      </label>
      <div
        className="field is-grouped"
      >
        <div
          className="control is-expanded"
        >
          <input
            className="input"
            disabled={true}
            readOnly={true}
            value=""
          />
        </div>
        <div
          className="control"
        >
          <button
            className="button"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            <span>
              Copy
            </span>
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
